library/bn_mul.h

```diff
@@ -673,6 +673,22 @@
 #define MULADDC_CANNOT_USE_R7
 #endif
 
+/*
+ * Similarly, we need to disable the assembly below if:
+ * - compiler is armclang
+ * - optimisation is not -O0
+ * - target is Thumb
+ * - target cpu is one of cortex-m0, cortex-m0plus, cortex-m1, cortex-m23, sc000
+ *
+ * Checking for __ARM_ARCH_6M__ or __ARM_ARCH_8M_BASE__ seems to identify exactly these
+ * cpus and no others (tested against all values for -mcpu known to armclang 6.20).
+ */
+#if defined(__ARMCC_VERSION) && defined(__OPTIMIZE__) && defined(__thumb__)
+#if defined(__ARM_ARCH_8M_BASE__) || defined(__ARM_ARCH_6M__)
+#define MULADDC_CANNOT_USE_R7
+#endif
+#endif
+
 #if defined(__arm__) && !defined(MULADDC_CANNOT_USE_R7)
 
 #if defined(__thumb__) && !defined(__thumb2__)
```

Comment:
So on which platforms will this condition be true? Testing with armclang 6.6, the architectures excluded by the new check are exactly the architectures where `__thumb__` is defined but not `__thumb2__`. So this block can now only be used by other compilers. These other compilers would have to define `__GNUC__` due to the surrounding condition `#if defined(__GNUC__) && \ ( !defined(__ARMCC_VERSION) || __ARMCC_VERSION >= 6000000 )`, i.e. they're GCC or a compiler such as Clang that masquerades as GCC.

But as I understand it, the problem is that we're trying to use instructions that don't exist for the target architecture, not that we're trying to use something that armclang doesn't support. So either the compilers that this block works with support an arm architecture that Armclang doesn't support (which seems weird), or this block actually doesn't work with other compilers, either.

Reply:
Hmm, I thought the list was different, but you're right, it's the same CPUs as per `defined(__thumb__) && !defined(__thumb2__)`. So the check could be simplified quite a bit.

There are a few problems, e.g.:

```
bignum_core.c:485:9: error: inline assembly requires more registers than available
        MULADDC_X1_INIT
        ^
./bn_mul.h:696:13: note: expanded from macro 'MULADDC_X1_INIT'
            "ldr    r0, %3                      \n\t"   \
```

and

```
error: invalid instruction, any one of the following would fix this:
<inline asm>:24:2: note: instantiated into assembly here
        lsr    r3, r10, #16
        ^
note: instruction requires: thumb2
```

and

```
note: too few operands for instruction
note: no flag-preserving variant of this instruction available
error: operand must be a register in range [r0, r7]
```

So it's more than just register selection I think.

However, you are right: gcc also fails to compile the assembly code when invoked with `make lib CFLAGS="-Os -g -mcpu=cortex-m0 -mthumb" CC=/usr/bin/arm-linux-gnueabi-gcc`

So I think this needs to be disabled for all compilers.

Reply:
Hmmm, I'm not very fluent in Arm assembly, but I wonder if the condition is simply swapped. Isn't the following block in fact Thumb2 code? Whereas the `#else` block below looks like Thumb (pure 16-bit) code. Notice the use of `add` with low registers in the first block, vs the use of `adds` with low registers in the else block.

It looks to me like (excluding the DSP case) we're using the Thumb code on platforms that support Thumb2 (successfully but presumably inefficiently), and the Thumb2 code on platforms that don't support Thumb2 (unsuccessfully).

Reply:
In summary, for CPUs where we have `defined(__thumb__) && !defined(__thumb2__)`: current behaviour is for `-O0`, we don't get assembly, and it works; or for `-O<non-zero>` we do get assembly, and it breaks. This holds for armclang and gcc.

I think the solution is simply to disable the asm in all cases where we have `defined(__thumb__) && !defined(__thumb2__)`

Reply:
But don't we in fact have assembly code for `defined(__thumb__) && !defined(__thumb2__)`? But in the `#else` part.

Reply:
Hmm, no, that's not it either: `umlal` is thumb2 only.

Reply:
The original asm would have been enabled only for the opposite case, where we have `defined(__thumb__) && !defined(__thumb2__)`, and not enabled if this wasn't the case. Maybe this was a typo and the intended logic was the opposite behaviour?

Reply:
Yes, I think the intended logic was
```
diff --git a/library/bn_mul.h b/library/bn_mul.h
index ab59fbd64f..8e1ccc824a 100644
--- a/library/bn_mul.h
+++ b/library/bn_mul.h
@@ -675,7 +675,7 @@
 
 #if defined(__arm__) && !defined(MULADDC_CANNOT_USE_R7)
 
-#if defined(__thumb__) && !defined(__thumb2__)
+#if !defined(__thumb__)
 
 #define MULADDC_X1_INIT                                 \
     asm(                                                \
@@ -796,7 +796,7 @@
         );                                                   \
     }
 
-#else
+#elif defined(__thumb2__)
 
 #define MULADDC_X1_INIT                                 \
     asm(                                                \
```

Reply:
The thumb code in the first block was added in 4f9a7bb7fdb52299c6b91b76252104a70d74710d with the condition `__thumb__`. No indication of what CPU was intended, but if Paul was aiming for A-class and not M-class, he might have treated “thumb” as synonym with “the thumb version that's available on A-class”, which is Thumb2. I don't know in what way the new code is preferable to the version before that commit though.

Reply:
There are three variants of the asm, call them A B and C (in the order they appear in the file). I've tested all the combinations, and found the following combinations work:

A works on:
Thumb, thumb2 gcc if not -O0
Arm

B works on:
Thumb2 gcc
Arm

C works on:
Thumb2 (except gcc -O0)
Arm

For "thumb" I've tested with cortex-m0 (armclang) and "-mcpu=arm1136jf-s -march=armv6j -mthumb" (gcc). For "thumb2", I'm using cortex-m33 for both. For arm, I used cortex-a78 on armclang, and "-march=armv7-a+simd -mfpu=crypto-neon-fp-armv8 -marm" for gcc.

Reply:
Ok, so I'm still confused about A. Why is it ok for gcc but not for armclang?

Reply:
Note that it's not enough to say a CPU has "Thumb-1" or "Thumb-2", you have to know which instruction encodings are actually available in its arch version (and the CPU could have more than the minimum in the Arch). Armv6-M and Armv8-M Baseline both have most (but not all) of the Thumb-1 encodings, including some ones that didn't exist in previous architectures, plus a tiny number of Thumb-2 encodings. Someone has tried to summarise all the Cortex-M variation on Wikipedia: https://en.wikipedia.org/wiki/ARM_Cortex-M#Instruction_sets.

Looking at the Armv6-M ARM, I can only find encodings for MOV with 3 bits for the destination register. So I think "mov    r9, r7" on line 687 and "mov    r8, r7" on line 690 might be the offending instructions for Armv6-M/Armv8-M-Baseline, needing 4 bits for their destination registers, even if a valid Thumb-1 encoding does exist for some other CPU.

I think generally it's better to gate ASM blocks like this with the Arch versions (__ARM_ARCH_8M_BASE__, etc) that it's known to be valid for, rather than use the thumb/thumb2 defs which cover quite a range of different instruction arch.

Does look like the change solves a longstanding issue, though I think it should be applied to v6-M/v8-M base generally rather than only on Arm Compiler. Also, I think it's essentially a coincidence that the def "MULADDC_CANNOT_USE_R7" compiles out the problematic assembly, rather than being anything to do with use of R7.

Hope that helps. ;)

Reply:
I don't know. armclang reports lots of things like:

```
bignum_core.c:485:9: error: inline assembly requires more registers than available
        MULADDC_X1_INIT
        ^
./bn_mul.h:705:13: note: expanded from macro 'MULADDC_X1_INIT'
            "ldr    r0, %3
	"               ^
error: invalid instruction, any one of the following would fix this:
<inline asm>:5:2: note: instantiated into assembly here
        lsr    r7, r3, #16
        ^
note: instruction requires: thumb2
<inline asm>:5:2: note: instantiated into assembly here
        lsr    r7, r3, #16
        ^
note: no flag-preserving variant of this instruction available
<inline asm>:5:2: note: instantiated into assembly here
        lsr    r7, r3, #16
        ^
note: too many operands for instruction
```

Reply:
Judging by that wikipedia page, it looks like __thumb2__ does exactly map to the CPUs which have the entire Thumb2 instruction set (which are the ones which need a different code path). So I think using __thumb2__ might actually be a more precise way to identify Thumb2 support than looking at `__ARM_ARCH_8M_BASE__` etc. @jf549 WDYT?

Reply:
Hmm, yeah there may be more wrong with it than just those two mov to high reg... I think you could use "lsrs" instead of "lsr" and this would be taking the "note: no flag-preserving variant of this instruction available" option of fixing the problem. But maybe it's important to preserve the flags here, I didn't check... But anyway, still the same point that some Thumb instructions written for one version of the Arch don't necessarily work on another.

Reply:
At least as it stands, yes there seems to be less variation with Thumb-2 support. You still rely on future Arch not removing the instruction encodings you currently use, but mostly that tends to be true. I suppose the problems here came about because Armv6-M was significantly cut down from the "A-class" version of Thumb this ASM was written for.